### ISSUE TITLE
Scheduler: Sort scheduling failure msgs to better analysis the reason why this node cannot afford pod

### ISSUE THREAD

**Issue by NoicFank** [kubernetes/kubernetes#123383]
### What would you like to be added?

Currently，the scheduling failure msgs from filter stage are as follows, where they are simply sorted by `sort.Strings()`.
```
0/8 nodes are available: 1 {out-of-tree-plugin-2 reason-2}, 2 Insufficient cpu, 2 {out-of-tree-plugin-1 reason-1}, 2 node(s) didn't match pod topology spread constraints, 3 Insufficient memory.
```

I want to sort the failure messages in the order of plugins in the filter stage of the profile.

Example: 
```
      - schedulerName: test-scheduler
        plugins:
          ...
          filter:
            enabled:
            - name: out-of-tree-plugin-1
            - name: out-of-tree-plugin-2
         ...
```

Then, the expected failure msgs are as follows:
```
0/8 nodes are available: 2 Insufficient cpu, 3 Insufficient memory, 2 node(s) didn't match pod topology spread constraints, 2 {out-of-tree-plugin-1 reason-1}, 1 {out-of-tree-plugin-2 reason-2}.
```

### Why is this needed?

After sorting the scheduling failure msgs according to the method proposed above, it is more convenient for us to investigate `the reason why a node cannot afford the pod accurately` without increasing the output information. 

Because we can determine the specific reason why the node cannot afford the pod one by one in the order of error messages, and we don't know which error message to start checking from before.

Example:

if nodeA can not pass the both `out-of-tree-plugin-2` and `NodeResourcesFit` plugins,  with the previous msgs:
```
0/8 nodes are available: 1 {out-of-tree-plugin-2 reason-2}, 2 Insufficient cpu, 2 {out-of-tree-plugin-1 reason-1}, 2 node(s) didn't match pod topology spread constraints, 3 Insufficient memory.
```
if we check the reason `{out-of-tree-plugin-2 reason-2}` before `Insufficient cpu` for nodeA (Due to we do not known which reason to check firstly), then we may mistakenly believe that the scheduling failure was caused by `reason-2`, rather than `Insufficient cpu`. While it actually caused by `Insufficient cpu` in this scheduling cycle.

After we using following proposed failure msgs:
```
0/8 nodes are available: 2 Insufficient cpu, 3 Insufficient memory, 2 node(s) didn't match pod topology spread constraints, 2 {out-of-tree-plugin-1 reason-1}, 1 {out-of-tree-plugin-2 reason-2}.
```
we can check failure reason  in the order mentioned above, and get the correct failure reason `Insufficient cpu` for nodeA in this scheduling cycle.

Besides, when we have this proposed order of error messages, we can also quickly identify the `failure reasonA` for the node, as it can avoid many unnecessary checks for the reason sorted after `failure reasonA`.

**Comment by NoicFank** [comment]
Besides, did we need a gate to print all the failure msgs with nodeName in log for easy debugging or analysis? While this gate defaults to false.

maybe like
```
"Unable to schedule pod; no fit; waiting" pod="xxx/xxx-4" err="0/9 nodes are available: 1 node(s) had taint {xxx} that the pod didn't tolerate [node1], 1 xxxx [node2], 2 Insufficient cpu [node3, node4], 2 xxx [node5, node6], 2 node(s) didn't match pod topology spread constraints [node7, node8], 3 Insufficient memory[node3, node4, node9]."
```

**Comment by sanposhiho** [comment]
> if nodeA can not pass the both out-of-tree-plugin-2 and NodeResourcesFit plugins, with the previous msgs:

If a node is rejected by one plugin, then the scheduler doesn't run all the other remaining plugins. How could nodeA have two failure plugins?
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework.go#L894

**Comment by NoicFank** [comment]
> > if nodeA can not pass the both out-of-tree-plugin-2 and NodeResourcesFit plugins, with the previous msgs:
> 
> If a node is rejected by one plugin, then the scheduler doesn't run all the other remaining plugins. How could nodeA have two failure plugins? https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework.go#L894

yes, one node only got one failure plugin reason in one scheduling cycle.

nodeA can not pass the both out-of-tree-plugin-2 and NodeResourcesFit plugins. This is `the fact`, Not the result of feedback from the scheduler.

the example above wants to show the awkward situation to quickly know the specified one failure reason for one node `in this scheduling cycle` if node actually do not satisfy multiple filters (because like you said, only one plugin will reject this pod in this scheduling cycle). 

After podA failed scheduling, we wan to check the failure reason for nodeA `in this scheduling cycle` if both failure reasons from  out-of-tree-plugin-2 and NodeResourcesFit plugins shows. (only one of them is the truly reason in this scheduling cycle)

Then , currently impl, we need to search which plugin run first to determine the specified failure reason for nodeA in this scheduling cycle. (maybe from kubeSchedulerConfiguration or even source code)

This proposal wants to reflect the execution order of filters plugins to scheduling event msg. Then we can simply get the specified failure reason for nodeA in this scheduling by the which failure reason comes first in the event msg.


Furthermore, we can quickly identify the true failure reason for node in this scheduling cycle by checking for failure reasons in order.

**Comment by sanposhiho** [comment]
> Then we can simply get the specified failure reason for nodeA in this scheduling by the which failure reason comes first in the event msg.

How? NodeA may or may not get rejected by other plugins. We cannot know until the scheduler runs plugins at scheduling retries. Also, either way, we don't represent the node name on the event message. (and we cannot; the event message has a small length limitation)

> we can quickly identify the true failure reason for node in this scheduling cycle by checking for failure reasons in order.

All the printed failure reasons are true failure reasons. It's just that they are not all of true failure reasons. (since, as you said, even if nodeA is rejected by plugin-a and plugin-b theoritically, the scheduler only shows the plugin name that has run first. )
